src/main/java/ru/curs/xylophone/MergeRegionContainer.java

```diff
@@ -0,0 +1,230 @@
+            mergedRegions.add(new MutableCellRangeAddress(res));
+        }
+    }
+
+    /**
+     * Apply all merged regions to sheet.
+     * @param sheet sheet of docs
+     */
+    public void apply(Sheet sheet) {
+        if (sheet == null) {
+            return;
+        }
+        mergedRegions.stream().map(MutableCellRangeAddress::toCellRangeAddress).forEach(sheet::addMergedRegion);
+        clear();
+    }
+
+    /**
+     * Find main region with current cell.
+     * For defining value in region.
+     * @param res range of current merge region.
+     * @return Current merge region if intersects is false; otherwise - intersected existing region.
+     */
+    public CellRangeAddress findIntersectedRange(CellRangeAddress res) {
+        if (res.getFirstColumn() < 0 || res.getFirstRow() < 0) {
+            throw new IllegalArgumentException("Cannot merge first row with upper cell or first column with left cell");
+        }
+        Optional<MutableCellRangeAddress> intersected = findIntersectedRegion(res);
+        if (intersected.isPresent()) {
+            return intersected.get().toCellRangeAddress();
+        } else {
+            return res;
+        }
+    }
+
+    /**
+     * Clear list.
+     */
+    private void clear() {
+        mergedRegions.clear();
+    }
+
+    /**
+     * Merge current cell with existing region.
+     * @param mergedRegion existing merged region.
+     * @param res current cell with it's neighbour (up or left)
+     */
+    private void mergeIntersectedRegions(MutableCellRangeAddress mergedRegion, CellRangeAddress res) {
+        if (isValidUnion(mergedRegion.toCellRangeAddress(), res)) {
+            CellRangeAddress unionRangeRes = new CellRangeAddress(mergedRegion.getFirstRow(), res.getLastRow(),
+                    mergedRegion.getFirstColumn(), res.getLastColumn());
+            mergedRegion.setAddress(unionRangeRes);
+        }
+    }
+
+    /**
+     * Check if is it valid union.
+     * @param mergedRegion existing merged region
+     * @param result region that will be concatenated.
+     * @return True if it is one row (or column); otherwise - false.
+     */
+    private boolean isValidUnion(CellRangeAddress mergedRegion, CellRangeAddress result) {
+        int startRow = mergedRegion.getFirstRow();
+        int startCol = mergedRegion.getFirstColumn();
+        int finishRow = mergedRegion.getLastRow();
+        int finishCol = mergedRegion.getLastColumn();
+
+        if (startRow == finishRow) {
+            return result.getFirstRow() == result.getLastRow();
+        }
+        return startCol == finishCol && result.getFirstColumn() == result.getLastColumn();
+    }
+
+    /**
+     * Find intersected existing region.
+     * @param res range address
+     * @return optional
+     */
+    private Optional<MutableCellRangeAddress> findIntersectedRegion(CellRangeAddress res) {
+        return mergedRegions
+                .stream()
+                .filter(addr -> addr.intersects(res))
```

Comment:
method reference should be used here

Reply:
AAAAAND hello, O(n^2) complexity! The more rectangles there are in the list, the more comparisons we are to make.
This **definitely** is going to cause problems for spreadsheets with large number of merged cells.
 
See https://stackoverflow.com/questions/13910287/data-structure-to-hold-list-of-rectangles